atom/browser/net/atom_network_delegate.cc

```diff
@@ -0,0 +1,388 @@
+  net::HttpRequestHeaders::Iterator it(headers);
+  while (it.GetNext())
+    header_dict->SetString(it.name(), it.value());
+  return header_dict;
+}
+
+base::DictionaryValue* GetResponseHeadersDict(
+    const net::HttpResponseHeaders* headers) {
+  base::DictionaryValue* header_dict = new base::DictionaryValue();
+  if (headers) {
+    void* iter = nullptr;
+    std::string key;
+    std::string value;
+    while (headers->EnumerateHeaderLines(&iter, &key, &value))
+      header_dict->SetString(key, value);
+  }
+  return header_dict;
+}
+
+void OnBeforeURLRequestResponse(
+    const net::CompletionCallback& callback,
+    GURL* new_url,
+    const AtomNetworkDelegate::BlockingResponse& result) {
+  if (!result.redirectURL.is_empty())
+    *new_url = result.redirectURL;
+  callback.Run(result.Cancel());
+}
+
+void OnBeforeSendHeadersResponse(
+    const net::CompletionCallback& callback,
+    net::HttpRequestHeaders* headers,
+    const AtomNetworkDelegate::BlockingResponse& result) {
+  if (!result.requestHeaders.IsEmpty())
+    *headers = result.requestHeaders;
+  callback.Run(result.Cancel());
+}
+
+void OnHeadersReceivedResponse(
+    const net::CompletionCallback& callback,
+    const net::HttpResponseHeaders* original_response_headers,
+    scoped_refptr<net::HttpResponseHeaders>* override_response_headers,
+    const AtomNetworkDelegate::BlockingResponse& result) {
+  if (result.responseHeaders.get()) {
+    *override_response_headers = new net::HttpResponseHeaders(
+        original_response_headers->raw_headers());
+    void* iter = nullptr;
+    std::string key;
+    std::string value;
+    while (result.responseHeaders->EnumerateHeaderLines(&iter, &key, &value)) {
+      (*override_response_headers)->RemoveHeader(key);
+      (*override_response_headers)->AddHeader(key + ": " + value);
+    }
+  }
+  callback.Run(result.Cancel());
+}
+
+}  // namespace
+
+// static
+std::map<AtomNetworkDelegate::EventTypes, AtomNetworkDelegate::ListenerInfo>
+AtomNetworkDelegate::event_listener_map_;
+
+AtomNetworkDelegate::AtomNetworkDelegate() {
+}
+
+AtomNetworkDelegate::~AtomNetworkDelegate() {
+}
+
+void AtomNetworkDelegate::SetListenerInIO(
+    EventTypes type,
+    scoped_ptr<base::DictionaryValue> filter,
+    const Listener& callback) {
+  ListenerInfo info;
+  info.callback = callback;
+
+  const base::ListValue* url_list = nullptr;
+  if (filter->GetList("urls", &url_list)) {
+    for (size_t i = 0; i < url_list->GetSize(); ++i) {
+      std::string url;
+      extensions::URLPattern pattern;
+      if (url_list->GetString(i, &url) &&
+          pattern.Parse(url) == extensions::URLPattern::PARSE_SUCCESS) {
+        info.url_patterns.insert(pattern);
+      }
+    }
+  }
+
+  event_listener_map_[type] = info;
+}
+
+int AtomNetworkDelegate::OnBeforeURLRequest(
+    net::URLRequest* request,
+    const net::CompletionCallback& callback,
+    GURL* new_url) {
+  brightray::NetworkDelegate::OnBeforeURLRequest(request, callback, new_url);
```

Comment:
We should check whether `kOnBeforeRequest` is in `event_listener_map_`, and if it is then execute code bellow, otherwise execute super class's method.

All overrides bellow should do the same.